src/ESM.t.sol

```diff
@@ -37,109 +59,169 @@ contract TestUsr {
     constructor(DSToken gem_) public {
         gem = gem_;
     }
-    function callFire(ESM esm) external {
-        esm.fire();
-    }
 
     function callJoin(ESM esm, uint256 wad) external {
         gem.approve(address(esm), uint256(-1));
 
         esm.join(wad);
     }
+
+    function callFire(ESM esm) external {
+        esm.fire();
+    }
+
+    function callBurn(ESM esm) external {
+        esm.burn();
+    }
+}
+
+contract Authority {
+  function canCall(address, address, bytes4 sig)
+      public view returns (bool)
+  {
+    if (sig == bytes4(0x42966c68)) { // burn(uint256)
+      return true;
+    } else {
+      return false;
+    }
+  }
 }
 
 contract ESMTest is DSTest {
     ESM     esm;
     DSToken gem;
+    address pauseProxy = address(123);
+    VatMock vat;
     EndMock end;
     uint256 min;
-    address pit;
     TestUsr usr;
     TestUsr gov;
 
     function setUp() public {
         gem = new DSToken("GOLD");
-        end = new EndMock();
+        gem.setAuthority(DSAuthority(address(new Authority())));
+        vat = new VatMock();
+        vat.rely(pauseProxy);
+        end = new EndMock(vat);
         usr = new TestUsr(gem);
-        gov = new TestUsr(gem);
-        pit = address(0x42);
     }
 
     function test_constructor() public {
-        esm = makeWithCap(10);
+        esm = new ESM(address(gem), address(end), pauseProxy, 10);
 
         assertEq(address(esm.gem()), address(gem));
         assertEq(address(esm.end()), address(end));
+        assertEq(address(esm.proxy()), pauseProxy);
         assertEq(esm.min(), 10);
-        assertEq(esm.fired(), 0);
+        assertEq(end.live(), 1);
     }
 
     function test_Sum_is_internal_balance() public {
-        esm = makeWithCap(10);
+        esm = new ESM(address(gem), address(end), address(0), 10);
         gem.mint(address(esm), 10);
 
         assertEq(esm.Sum(), 0);
     }
 
-    function test_fire() public {
-        esm = makeWithCap(0);
-        gov.callFire(esm);
+    function test_fire_deny_proxy() public {
+        esm = new ESM(address(gem), address(end), pauseProxy, 0);
+        vat.rely(address(esm));
+        assertEq(vat.wards(pauseProxy), 1);
+        usr.callFire(esm);
+        assertEq(vat.wards(pauseProxy), 0);
+
+        assertEq(end.live(), 0);
+    }
+
+    function test_fire_no_proxy_action() public {
+        esm = new ESM(address(gem), address(end), address(0), 0);
+        assertEq(vat.wards(pauseProxy), 1);
+        usr.callFire(esm);
+        assertEq(vat.wards(pauseProxy), 1);
 
-        assertEq(esm.fired(), 1);
         assertEq(end.live(), 0);
     }
 
     function testFail_fire_twice() public {
```

Comment:
Not sure that this test makes any sense to keep now--it is really just testing behavior of `EndMock`, the ESM no longer has any logic to prevent calling `fire` twice. So it's a bit misleading I think, if we believe that tests should somehow document the intent of the code.

Reply:
Wait, ignore, I'm reviewing code too early in the morning again...

Reply:
haha